### PR TITLE
Clean up Webpack configuration

### PR DESCRIPTION
**Pull request by maximemulder**
## Brief summary of changes

This is the first of a series of PRs that will optimize the Webpack configuration. This PR only cleans up the current configuration with more explicit typing, and isolating some logic into functions.

The code is sometimes a little heavy because I use a rather functional approach, but I believe this leads to a cleaner logic (create new record instead of mutating existing one).

## Testing

```
(loris-mri-python) lorisadmin@mmulder-dev:/var/www/loris$ time npm run compile

> loris@1.0.0 compile
> webpack --node-env=development

Building all modules
webpack compiled successfully

webpack compiled successfully

real	2m29.071s
user	1m42.369s
sys	0m5.481s
(loris-mri-python) lorisadmin@mmulder-dev:/var/www/loris$ export target=login
(loris-mri-python) lorisadmin@mmulder-dev:/var/www/loris$ time npm run compile

> loris@1.0.0 compile
> webpack --node-env=development

Building module 'login'
webpack compiled successfully

real	0m13.322s
user	0m11.475s
sys	0m1.019s
```

##  Oh nooo, static tests are failing 😱😱😱

I am on static tests strike until #9437 is merged. EDIT: Yoohoo, it got merged and now the tests work :partying_face:

### CODE REVIEW

File: webpack.config.ts [hunk 1]
```diff
@@ -186,35 +192,18 @@ const plugins: webpack.WebpackPluginInstance[] = [];
 plugins.push(new CopyPlugin({
   patterns: [
     {
-      from: path.resolve(__dirname, 'node_modules/react/umd'),
-      to: path.resolve(__dirname, 'htdocs/vendor/js/react'),
+      from: `node_modules/react/umd/${
```
Comment (skarya22) [inline webpack.config.ts:195]: i like this change!

File: webpack.config.ts [hunk 2]
```diff
@@ -186,35 +192,18 @@ const plugins: webpack.WebpackPluginInstance[] = [];
...
+        isDev ? 'react.development.js' : 'react.production.min.js'
+      }`,
+      to: 'htdocs/vendor/js/react',
       force: true,
-      globOptions: {
-        ignore: ['react.profiling.min.js'],
-      },
```
Comment (skarya22) [inline webpack.config.ts:194]: Why was this removed / what is the implication of its removal?

File: webpack.config.ts [hunk 3]
```diff
@@ -232,63 +221,110 @@ if (EEGVisEnabled !== 'true' && EEGVisEnabled !== '1' ) {
...
+      },
+    },
+  }));
+}
+
+/**
+ * Get the Webpack entries of the LORIS modules to build, with each entry being
+ * mapped to its name.
+ */
+function getModulesEntries(): Record<string, webpack.EntryOptions> {
```
Comment (skarya22) [inline webpack.config.ts:317]: I think the return type of Object.assign is not the same as the return type mentioned up here

**Review by skarya22 — CHANGES_REQUESTED** [review]

File: webpack.config.ts (hunk 2 above)
Comment (maximemulder) [reply webpack.config.ts:194]: TBH I am pretty sure it was dead code. Before, all the whole `umd` directory was marked for copy, but the `filter` function ignored all files but two (`react.development.js` and `react.production.min.js`). Now, only these two files are copied (or only the one needed depending on whether this is a dev or prod build more precisely), so there is no need to ignore the other files of the directory anymore.

File: webpack.config.ts (hunk 1 above)
Comment (maximemulder) [reply webpack.config.ts:195]: Yup, I had to look at the documentation to see if non-absolute paths were ok !

File: webpack.config.ts (hunk 3 above)
Comment (maximemulder) [reply webpack.config.ts:317]: It is ! `moduleEntries` is of type `Record<string, webpack.EntryOptions>[]`. `Object.assign` is used with the spread of `moduleEntries` to merge all of these objects into a single object. The Typescript compiler does type checking here, so it would have signaled an error if the type did not match.

image

Comment (skarya22) [reply webpack.config.ts:317]: Hmm, I see that the type of the returned value is `Record<string, webpack.EntryOptions>[]`, but doesn't the return type of the function say `Record<string, webpack.EntryOptions>`? With no `[]`?

Comment (skarya22) [reply webpack.config.ts:317]: ohhhhh I see - thanks!
